File: java-sdk/src/main/java/com/cleanster/sdk/api/BlacklistApi.java
package com.cleanster.sdk.api;

import com.cleanster.sdk.client.HttpClient;
import com.cleanster.sdk.model.*;
import com.fasterxml.jackson.core.type.TypeReference;

/**
 * API operations for the cleaner blacklist: list, add, and remove blacklisted cleaners.
 */
public class BlacklistApi {

    private final HttpClient httpClient;

    public BlacklistApi(HttpClient httpClient) {
        this.httpClient = httpClient;
    }

    /**
     * List all blacklisted cleaners.
     *
     * @return API response with list of blacklisted cleaners
     */
    public ApiResponse<Object> listBlacklistedCleaners() {
        return httpClient.get("/v1/blacklist/cleaner",
                new TypeReference<ApiResponse<Object>>() {});
    }

    /**
     * Add a cleaner to the blacklist.
     *
     * @param request Cleaner ID and reason
     * @return API response
     */
    public ApiResponse<Object> addToBlacklist(BlacklistRequest request) {
        return httpClient.post("/v1/blacklist/cleaner", request,
                new TypeReference<ApiResponse<Object>>() {});
    }

    /**
     * Remove a cleaner from the blacklist.
     *
     * @param request Cleaner ID to remove
     * @return API response
     */
    public ApiResponse<Object> removeFromBlacklist(BlacklistRequest request) {
        return httpClient.delete("/v1/blacklist/cleaner", request,
                new TypeReference<ApiResponse<Object>>() {});
    }
}

File: android-sdk/src/main/kotlin/com/cleanster/android/api/BlacklistApi.kt
package com.cleanster.android.api

import com.cleanster.android.model.*
import retrofit2.Retrofit
import retrofit2.http.*

internal interface BlacklistService {
    @GET("v1/blacklist/cleaner")
    suspend fun listBlacklistedCleaners(): ApiResponse<List<BlacklistEntry>>

    @POST("v1/blacklist/cleaner")
    suspend fun addToBlacklist(@Body body: AddToBlacklistRequest): ApiResponse<Any>

    @HTTP(method = "DELETE", path = "v1/blacklist/cleaner", hasBody = true)
    suspend fun removeFromBlacklist(@Body body: RemoveFromBlacklistRequest): ApiResponse<Any>
}

class BlacklistApi(retrofit: Retrofit) {
    private val service = retrofit.create(BlacklistService::class.java)

    suspend fun listBlacklistedCleaners() = wrap { service.listBlacklistedCleaners() }

    suspend fun addToBlacklist(cleanerId: Int, reason: String? = null) =
        wrap { service.addToBlacklist(AddToBlacklistRequest(cleanerId, reason)) }

    suspend fun removeFromBlacklist(cleanerId: Int) =
        wrap { service.removeFromBlacklist(RemoveFromBlacklistRequest(cleanerId)) }
}
